src/bigBed.c

```diff
@@ -0,0 +1,317 @@
+  struct bigBedInterval *hits = NULL, *tail = NULL;
+  /* querying records in range */
+  for (int i = 0; i < n_ranges; ++i) {
+    struct bigBedInterval *queryHits =
+      bigBedIntervalQuery(file, (char *)CHAR(STRING_ELT(r_seqnames, i)),
+                          start[i] - 1, start[i] - 1 + width[i], 0, lm);
+    if (!hits) {
+      hits = queryHits;
+      tail = slLastEl(hits);
+    } else {
+      tail->next = queryHits;
+      tail = slLastEl(tail);
+    }
+    INTEGER(n_qhits)[i] = slCount(queryHits);
+  }
+
+  /* need these before closing file */
+  char *asText = bigBedAutoSqlText(file);
+  struct asObject *as = asParseText(asText);
+  freeMem(asText);
+  int fieldCount = file->fieldCount;
+  int definedFieldCount = getDefinedFieldCount(as);
+  int extraFieldCount = fieldCount - definedFieldCount;
+  int n_hits = slCount(hits);
+  bigBedFileClose(&file);
+
+  int presentFieldCount = 0, unprotectCount = 0;
+  /* mandatory default field */
+  chromStart = PROTECT(allocVector(INTSXP, n_hits));
+  chromWidth = PROTECT(allocVector(INTSXP, n_hits));
+  /* if any of the default field is present and selected, allocate memory for it */
+  if (isPresent(definedFieldCount, i_name) && isSelected(r_defaultindex, 1)) {
+    name = PROTECT(allocVector(STRSXP, n_hits));
+    ++presentFieldCount;
+    ++unprotectCount;
+  }
+  if (isPresent(definedFieldCount, i_score) && isSelected(r_defaultindex, 2)) {
+    score = PROTECT(allocVector(INTSXP, n_hits));
+    ++presentFieldCount;
+    ++unprotectCount;
+  }
+  if (isPresent(definedFieldCount, i_strand)) {
+    strand = PROTECT(allocVector(STRSXP, n_hits));
+    ++unprotectCount;
+  }
+  if (isPresent(definedFieldCount, i_thick) && isSelected(r_defaultindex, 3)) {
+    thickStart = PROTECT(allocVector(INTSXP, n_hits));
+    thickWidth = PROTECT(allocVector(INTSXP, n_hits));
+    ++presentFieldCount;
+    unprotectCount += 2;
+  }
+  if (isPresent(definedFieldCount, i_itemRgb) && isSelected(r_defaultindex, 4)) {
+    itemRgb = PROTECT(allocVector(STRSXP, n_hits));
+    ++presentFieldCount;
+    ++unprotectCount;
+  }
+  if (isPresent(definedFieldCount, i_blocks) && isSelected(r_defaultindex, 5)) {
+      blocks = PROTECT(allocVector(VECSXP, n_hits));
+      ++presentFieldCount;
+      ++unprotectCount;
+  }
+
+  SEXPTYPE *typeId;
+  /* if extra fields are present and selected
+   * identify the type information and allocate memory */
+  if (extraFieldCount > 0) {
+    int k = 0, i = 0;
+    enum asTypes fieldType;
+    struct asColumn *asCol = as->columnList;
+    extraFields = PROTECT(allocVector(VECSXP, extraFieldCount));
+    typeId = (SEXPTYPE*)R_alloc(extraFieldCount, sizeof(SEXPTYPE));
+    for (int j = 0; j < fieldCount; ++j) {
+      fieldType = asCol->lowType->type;
+      if (j >= definedFieldCount) {
+        if (asTypesIsFloating(fieldType) || fieldType == t_int ||
+            fieldType == t_short || fieldType == t_byte || fieldType == t_off) {
+          typeId[k] = REALSXP;
```

Comment:
It looks like `t_short` and `t_byte` could fit into `INTSXP`. As long as the size is <= 32, it should fit an R integer vector. Here is the mapping:

```
  t_double,   /* REALSXP */
    t_float,    /* REALSXP */
    t_char,     /* STRSXP */
    t_int,      /* INTSXP */
    t_uint,     /* REALSXP because R's integers are signed */
    t_short,    /* INTSXP */
    t_ushort,   /* INTSXP */
    t_byte,     /* INTSXP */
    t_ubyte,    /* RAWSXP */
    t_off,      /* REALSXP although not ideal because double only has 52 bits of integer range */
    t_string,   /* STRSXP */
```